// @mischnic/async-hooks/lib/main/AsyncHook.js
"use strict";
exports.__esModule = true;
var AsyncHook = (function () {
    function AsyncHook(callbacks) {
        this.enabled = false;
        this.callbacks = callbacks;
    }
    AsyncHook.prototype.init = function (asyncId, type, triggerAsyncId, resource) {
        if (this.enabled && this.callbacks.init) {
            this.callbacks.init(asyncId, type, triggerAsyncId, resource);
        }
    };
    AsyncHook.prototype.before = function (asyncId) {
        if (this.enabled && this.callbacks.before) {
            this.callbacks.before(asyncId);
        }
    };
    AsyncHook.prototype.after = function (asyncId) {
        if (this.enabled && this.callbacks.after) {
            this.callbacks.after(asyncId);
        }
    };
    AsyncHook.prototype.destroy = function (asyncId) {
        if (this.enabled && this.callbacks.destroy) {
            this.callbacks.destroy(asyncId);
        }
    };
    AsyncHook.prototype.enable = function () {
        this.enabled = true;
        return this;
    };
    AsyncHook.prototype.disable = function () {
        this.enabled = false;
        return this;
    };
    return AsyncHook;
}());
exports.AsyncHook = AsyncHook;
