// @mischnic/async-hooks/lib/main/AsyncHooks.js
"use strict";
exports.__esModule = true;
var async_hooks_1 = require("./async-hooks");
var async_wrap_1 = require("./async-wrap");
var AsyncHook_1 = require("./AsyncHook");
var Hooks_1 = require("./Hooks");
var State_1 = require("./State");
var utils_1 = require("./utils");
var state = new State_1.State();
var hooks = Hooks_1.createHooks(state);
exports.version = require('../../package.json').version;
if (utils_1.packageExists('async_hooks')) {
    async_hooks_1.startAsyncHooks(hooks, state);
}
else {
    async_wrap_1.startAsyncWrap(hooks, state);
}
function createHook(callbacks) {
    var hook = new AsyncHook_1.AsyncHook(callbacks);
    hooks.add(hook);
    return hook;
}
exports.createHook = createHook;
function executionAsyncId() {
    return state.currentId;
}
exports.executionAsyncId = executionAsyncId;
function currentId() {
    console.warn('DeprecationWarning: async_hooks.currentId is deprecated. Use async_hooks.executionAsyncId instead');
    return executionAsyncId();
}
exports.currentId = currentId;
function triggerAsyncId() {
    return state.parentId;
}
exports.triggerAsyncId = triggerAsyncId;
function triggerId() {
    console.warn('DeprecationWarning: async_hooks.triggerId is deprecated. Use async_hooks.triggerAsyncId instead');
    return triggerAsyncId();
}
exports.triggerId = triggerId;
